ob1k-cache/src/main/java/com/outbrain/ob1k/cache/metrics/MonitoredTypedCache.java

```diff
@@ -0,0 +1,119 @@
+package com.outbrain.ob1k.cache.metrics;
+
+import java.util.Map;
+import com.google.common.base.Preconditions;
+import com.outbrain.ob1k.cache.EntryMapper;
+import com.outbrain.ob1k.cache.TypedCache;
+import com.outbrain.ob1k.concurrent.ComposableFuture;
+import com.outbrain.ob1k.concurrent.Consumer;
+import com.outbrain.ob1k.concurrent.Try;
+import com.outbrain.swinfra.metrics.api.MetricFactory;
+import com.outbrain.swinfra.metrics.api.Timer;
+
+/**
+ * A decorator adding metric monitoring to TypedCache.
+ *
+ * @author hunchback
+ */
+public class MonitoredTypedCache<K, V> implements TypedCache<K, V> {
+  protected final TypedCache<K, V> inner;
+
+  private final AsyncGetMetrics getAsyncMetrics;
+  private final AsyncBulkGetMetrics getBulkAsyncMetric;
+  private final AsyncOperationMetrics setAsyncMetrics;
+  private final AsyncOperationMetrics setBulkAsyncMetrics;
+  private final AsyncOperationMetrics deleteAsyncMetrics;
+
+  public MonitoredTypedCache(final TypedCache<K, V> inner, String cacheName, final MetricFactory metricFactory) {
+    Preconditions.checkNotNull(inner);
+    this.inner = inner;
+    Preconditions.checkNotNull(metricFactory);
+    final String component = inner.getClass().getSimpleName() + "." + cacheName;
+
+    getAsyncMetrics = new AsyncGetMetrics(metricFactory, component + ".getAsync");
```

Comment:
> hits / totals should be unified for get and getBulk. It doesn't matter how you call it, the keys are the same

one option (for eliminating the hit/total duplication) would be to unify get/getBulk into a single "fat" class: but then we would not override the finished(...) but would rather add new methods finishedGet(...) and finishedGetBulk(...) and then we are still left with a dangling finished(...). So I am more inclined to go for the more OO alternative of Base (hit/total are protected) which serves as a base for the AsyncGetMetrics and AsyncBulkGetMetrics ... a lot more classes but would make logic much more straight forward.